frontend/src/cljs/imas_seamap/map/events.cljs

```diff
@@ -159,32 +150,51 @@
+    
+    (let [responses     (get-in db [:feature-query :responses])
+          feature-infos (map response-to-info responses)
+          feature-info  (reduce combine-feature-info {} feature-infos)
+          had-insecure? (get-in db [:feature-query :had-insecure?])
+          candidate     (merge
+                         {:location point
+                          :had-insecure? had-insecure?}
+                         feature-info)]
+      (assoc db :feature candidate))))
+
+(defn got-feature-info [db [_ request-id point info-format response]]
   (if (not= request-id (get-in db [:feature-query :request-id]))
-    db                           ; Ignore late responses to old clicks
-    (let [feature-info (case info-format
-                         "text/html" (feature-info-html response)
-                         "application/json" (feature-info-json response)
-                         (feature-info-none))
-          db' (update-in db [:feature-query :response-remain] dec)
-          {:keys [response-priority response-remain candidate had-insecure?]} (:feature-query db')
-          higher-priority? (< priority response-priority)
-          candidate' (merge {:location point :had-insecure? had-insecure?} feature-info)]
-      (cond-> db'
-          ;; If this response has a higher priority, update the response candidate
-        higher-priority?
-        (assoc-in [:feature-query :candidate] candidate')
-          ;; If this is the last response expected, update the displayed feature
-        (zero? response-remain)
-        (assoc :feature (if higher-priority? candidate' candidate))))))
-
-(defn got-feature-info-error [db [_ _request_id priority _]]
-  (let [db (update-in db [:feature-query :response-remain] dec)
-        {:keys [response-priority response-remain]} (:feature-query db)
-        higher-priority? (< priority response-priority)]
-    (if (and (zero? response-remain) higher-priority?)
-      (assoc-in db [:feature :status] :feature-info/error)
-      db)))
+    db ; Ignore late responses to old clicks
+    (-> db
+        (update-in [:feature-query :response-remain] dec)
+        (update-in [:feature-query :responses] conj {:response response :info-format info-format})
+        (cond->
+          (zero? (dec (get-in db [:feature-query :response-remain])))
+          (display-feature-info point))))) ;; If this is the last response expected, update the displayed feature
+
+(defn got-feature-info-error [db [_ request-id point _]]
+  (if (not= request-id (get-in db [:feature-query :request-id]))
+    db ; Ignore late responses to old clicks
+    (-> db
+        (update-in [:feature-query :response-remain] dec)
+        (update-in [:feature-query :responses] conj nil)
+        (cond->
+         (zero? (dec (get-in db [:feature-query :response-remain])))
+          (display-feature-info point))))) ;; If this is the last response expected, update the displayed feature
```

Comment:
This is the only bit that looks a bit suss now; I think it's just because I'm used to seeing `update-in` and the like in these chains, where you're explicit about what's being updated, whereas this is a bit opaque.  I don't immediately see a better way though.  Perhaps it's the name?  (I guess you're not really displaying it here, you're formatting for display)

(Also, is that an indenting issue I just noticed?)

Reply:
Hmm, there's a single update at the end; perhaps it could instead return the formatted response, and got-feature-info assocs that on instead?  That feels like a better separation of concerns to me.

Reply:
Sounds good

Reply:
I've restructured the code to just return the summarised response, as suggested.